### ISSUE TITLE
ConfigDataLoader logging

### ISSUE THREAD

**Issue by wind57** [spring-cloud/spring-cloud-kubernetes#1173]
I was looking at an issue that a user reported, and while debugging it saw an interesting thing. 

In `KubernetesConfigDataLoader`, if we added some logging, for example:

```
private final Log log;

	public KubernetesConfigDataLoader(DeferredLogFactory factory) {
		log = factory.getLog(getClass());
	}

	@Override
	public ConfigData load(ConfigDataLoaderContext context, KubernetesConfigDataResource resource)
			throws IOException, ConfigDataResourceNotFoundException {

		log.debug("Kicking off ConfigDataLoader processing");

                ....

```

`DeferredLogFactory` is injected and I do see the logs I generate. But then, this method calls:

`propertySources.add(bootstrapContext.get(ConfigMapPropertySourceLocator.class).locate(env));` or `ConfigMapPropertySourceLocator::locate` and the logs there are not present. I've tried to play around with `DeferredLog`, but with no luck. 

I also saw [this issue](https://github.com/spring-projects/spring-boot/issues/24988), but again, did not either understand or did not implement correctly. 

Any hints into the direction of how to log these would be lovely, if any is possible at all. 

Without this, we are kind of blind what is going on

**Comment by ryanjbaxter** [comment]
@philwebb do you have any insights?

**Comment by philwebb** [comment]
The `DeferredLogFactory` instance passed to the `KubernetesConfigDataLoader` constructor should be transferred to a real logger once all environment processing has occurred.

The exact order should be:
* `EnvironmentPostProcessorApplicationListener` is created to trigger `Environment` post-processing. It includes a `DeferredLogs` instance (which is a `DeferredLogFactory`) that gets past to post-processors.
* The `ConfigDataEnvironmentPostProcessor` is one such post-processor. It will create `ConfigDataLoader` instances and should pass the same `DeferredLogs` instance along.
* The loader can call `DeferredLogFactory.getLog(...)` to create a `Log` instance. Calls on the resulting logger are tracked but not actually logged at this point.
* After all post-processing has finished, the `EnvironmentPostProcessorApplicationListener.finish()` method should be called.
* This triggers `DeferredLogs.switchOverAll()` which should finally push the log messages to a correctly configured logger.

Adding a breakpoint to `DeferredLogs.switchOverAll()` would be a good place to start debugging why you're not actually seeing log messages.

**Comment by wind57** [comment]
thank you @philwebb , but it seems to me that we are talking about slightly different issues here.

- I _do_ see logs from `KubernetesConfigDataLoader` - no problem here.
- The problem is that `KubernetesConfigDataLoader::load` internally calls `ConfigMapPropertySourceLocator::locate` and that one does some logging also. This logs is what I do not see, and that is the problem.

This is the chain that I was wondering if it's possible to make it work. I hope I make sense.

**Comment by wind57** [comment]
If I look at this class [here](https://github.com/spring-projects/spring-vault/commit/8b7f7e3a9b2d2a49195c5785845363d96463c233#diff-a69f0f13c2bcf444ea1a849ad05979ca71243e6f9249df4d694049732e27e95d), I notice : 

```
@SuppressWarnings("FieldMayBeFinal") // allow setting via reflection.
private static Log logger = LogFactory.getLog(LeaseAwareVaultPropertySource.class);
```

and in [this usage](https://github.com/spring-cloud/spring-cloud-vault/blob/main/spring-cloud-vault-config/src/main/java/org/springframework/cloud/vault/config/VaultConfigDataLoader.java#L109) here

I notice that they override the loggers, otherwise no logging is going to work - just like in our case.

Ryan, it seems to me that we will need to do the same thing

**Comment by wind57** [comment]
If you agree with the above, I have a PR [here](https://github.com/spring-cloud/spring-cloud-kubernetes/pull/1174/files) for fixing it.

**Comment by ryanjbaxter** [comment]
@wind57 have you tried passing the `DeferredLogFactory` to `ConfigMapPropertySourceLocator`?  Could we have a setter that if called overrides the default Logger?

**Comment by wind57** [comment]
we could, absolutely Ryan. But then there are many other classes that are called internally... which makes this setting impossible. We would hav to pass this logger everywhere.

I also agree that it's a bit hacky to do it with reflection, but if Vault does it...

**Comment by wind57** [comment]
either way, we need a solution here and fast (imho), because users using config-data are totally without logs, not being able to tell if anything is wrong at all.

**Comment by wind57** [comment]
although I gave up on this idea, I will expres it still: we can use `ThreadLocal` and set the logger into it and have a "proxy" that will check if the logger is present in ThreadLocal. The only reason I did not even try to implement something like this, is the fact that vault was using reflection and the be fair, I thought that this is OK and may be we want to be consistent. may be I was wrong.

**Comment by ryanjbaxter** [comment]
> we could, absolutely Ryan. But then there are many other classes that are called internally... which makes this setting impossible. We would have to pass this logger everywhere.

I dont think so, we leave the creation of the logger as is just not making it `static` then we add a setter and just call the setter in the config data case.

**Comment by wind57** [comment]
OK, so you are saying that `ConfigMapPropertySourceLocator` should have a setter where we can pass a different logger and override the existing one. So far, yes, it makes sense.

But `ConfigMapPropertySourceLocator::locate` calls internally `KubernetesClientConfigMapPropertySourceLocator::getMapPropertySource` that calls internally `KubernetesClientUtils::getApplicationNamespace`. How is the logger going to be passed down to `KubernetesClientUtils`? what am I missing?

**Comment by ryanjbaxter** [comment]
Yeah, I guess we would end up passing the logger around.  But wouldn't that be the same problem with reflections where we would have to be calling all these classes and setting the logger in them?

**Comment by wind57** [comment]
imho, reflection is easier to manage, and no obscure code in the logic of how things are really done. There would be additions of public methods for those setters.

It's a lot easier to manage too, if u ask me.

**Comment by ryanjbaxter** [comment]
@philwebb what do you think?  I like you don't think reflection is the right approach, but we can't be the only project (in fact Vault also is suffering from this) that would have the pass the Logger around in order to get things working.

**Comment by wind57** [comment]
We should replace the question label with a bug label

**Comment by wind57** [comment]
I was also wondering, considering the timelines for the next release (which I don't know what is), what if we have some solution in (even the reflection one), but keep this issue open in case we come up with something smarter? I feel weird for the fact that we made some good work on that config-data functionality and it's kind of useless in case things don't work :(

**Comment by ryanjbaxter** [comment]
The release will happen either tomorrow or today.  I would rather not rush.  The feature still works, its just logging that is the problem so this is not a blocker IMO.  A service release will certainly be planned shortly after because as always we will have issues that come up.

**Comment by philwebb** [comment]
>  I like you don't think reflection is the right approach, but we can't be the only project (in fact Vault also is suffering from this) that would have the pass the Logger around in order to get things working.

No, it's unfortunately a bit of a tricky issue and I don't really have a good suggestion. We have https://github.com/spring-projects/spring-boot/issues/28678 open in Spring Boot to look at it generally, but we're unlikely to get to it anytime soon.

**Comment by wind57** [comment]
seems to me like we won't have an option

**Comment by ryanjbaxter** [comment]
Thanks @philwebb.  

@wind57 lets tackle this after the new year.  I am on vacation beginning next week and wont be back till after the new year.  I will spend some time when I get back looking at the options.

**Comment by wind57** [comment]
of course! happy Holidays in such a case Ryan!

**Comment by ryanjbaxter** [comment]
you too!

**Comment by wind57** [comment]
I have two proposals of implementations to may be solve this:

[one here](https://github.com/spring-cloud/spring-cloud-kubernetes/pull/1190/files) that uses the new feature of JDK : hidden classes.

[another one here](https://github.com/spring-cloud/spring-cloud-kubernetes/pull/1174), that just uses reflection.

**Comment by ryanjbaxter** [comment]
Thanks!

I guess I am a little hesitant to make big changes, to just rip them all out when Boot fixes the issue.

**Comment by wind57** [comment]
_If_ they fix the issue, as @philwebb says there are no plans for this one in the near future. Personally, I see no problems with either hidden classes nor reflection, as long as we provide logging for the end-users.

To ease the transition in the future, I wanted to:

- create a PR with tests that assert logging, so that we have tests and ensure the build fails if something is not OK
- have a separate PR that drops this implementation (whatever we choose from the two), so that is eases our future Boot integration, when such is available. 

I feel somehow awkward having even a band-aid, but then not providing it... just my 0.02$

**Comment by ryanjbaxter** [comment]
@philwebb does this still seem like its going to happen in the near future?

**Comment by wind57** [comment]
I sound like a broken record here, still, what is a path forward here? If any. I am not trying to be pushy at all, it's just that this issue stays like a knot in my throat after all the pain I had to go with the user that reported it, nothing else :) thank you upfront.

**Comment by ryanjbaxter** [comment]
@wind57 please keep bugging me, I need to get with @philwebb to have a quick discussion
I have a couple of other things on my plate at the moment but I will try to get to this soon.

**Comment by wind57** [comment]
@ryanjbaxter random 2 weeks reminder... :)

**Comment by ryanjbaxter** [comment]
I have discussed this issue with a number of people on the team, ultimately everyone has come back to the point that the amount of code needed to fix the issue outweighs the issue itself.  We would be much more concerned if this caused the application to not work properly, but ultimately the app will work fine, we just might run into some trouble debugging things.  

I think at this point we close this issue, as its something Spring Boot needs to address.

I also want to acknowledge your effort in trying to find a viable solution to the problem, that has not gone unnoticed 🙌

**Comment by wind57** [comment]
oh it was a fun issue digging in! I just feel already bad for whomever is going to triage an issue when it comes (like the one we had) and logs are mute :) but I can't disagree either... Closing. 

I am also going to close the two PRs I had with proposals. 
In the background I am still going to maintain one of them, in case a future issue comes in and I want to help

**Comment by kguelzau** [comment]
So there are log statements but the output is actually never logged due to https://github.com/spring-projects/spring-boot/issues/28678 right?
Is there any workaround to see the logoutput to find erros e.g. missing permissions for the service account to load a config map?
